test/validation/api/scheduler/scheduler.c

```diff
@@ -2351,6 +2401,78 @@ static void scheduler_test_pause_enqueue(void)
 	CU_ASSERT(odp_queue_destroy(queue) == 0);
 }
 
+static void scheduler_test_private_groups(void)
+{
+	odp_schedule_capability_t sched_capa;
+	int num_private_groups, num_thr = globals->num_workers;
+	odp_pool_t pool = odp_pool_lookup(MSG_POOL_NAME);
+	odp_thrmask_t empty_mask;
+	odp_queue_param_t queue_param;
+	odp_buffer_t buffer;
+	char group_name[ODP_SCHED_GROUP_NAME_LEN];
+	char queue_name[ODP_QUEUE_NAME_LEN];
+	int idx[MAX_WORKERS];
+	void *args[MAX_WORKERS];
+	uint32_t num_predef = 3; /* Predefined groups */
+
+	CU_ASSERT_FATAL(globals != NULL);
+	CU_ASSERT_FATAL(!odp_schedule_capability(&sched_capa));
+	CU_ASSERT_FATAL(pool != ODP_POOL_INVALID);
+
+	num_private_groups = (sched_capa.max_groups > num_predef) ?
+			     (sched_capa.max_groups - num_predef) : 0;
+
+	if (num_private_groups == 0) {
+		printf("\nNot enough schedule groups. %s skipped.\n", __func__);
+		return;
+	}
+
+	if (num_thr > num_private_groups)
+		num_thr = num_private_groups;
+
+	odp_barrier_init(&globals->barrier, num_thr);
+	odp_thrmask_zero(&empty_mask);
+
+	for (int i = 0; i < num_thr; i++) {
+		snprintf(group_name, sizeof(group_name), "pgt_group_%d", i);
+		snprintf(queue_name, sizeof(queue_name), "pgt_queue_%d", i);
+		globals->pgt_groups[i] = odp_schedule_group_create(group_name, &empty_mask);
+		CU_ASSERT_FATAL(globals->pgt_groups[i] != ODP_SCHED_GROUP_INVALID);
+
+		odp_queue_param_init(&queue_param);
+		queue_param.type        = ODP_QUEUE_TYPE_SCHED;
+		queue_param.sched.sync  = ODP_SCHED_SYNC_PARALLEL;
+		queue_param.sched.group = globals->pgt_groups[i];
+		queue_param.size        = PGT_NUM_EV;
+
+		globals->pgt_queues[i] = odp_queue_create(queue_name, &queue_param);
+		CU_ASSERT_FATAL(globals->pgt_queues[i] != ODP_QUEUE_INVALID);
+
+		for (uint32_t j = 0; j < PGT_NUM_EV; j++) {
+			buffer = odp_buffer_alloc(pool);
+			CU_ASSERT_FATAL(buffer != ODP_BUFFER_INVALID);
+			CU_ASSERT_FATAL(odp_queue_enq(globals->pgt_queues[i],
+						      odp_buffer_to_event(buffer)) == 0);
+		}
+	}
+
+	for (int i = 0; i < num_thr; i++) {
+		idx[i] = i;
+		args[i] = &idx[i];
+	}
```

Comment:
Perhaps it would be a bit simpler to use an `uintptr_t` array (see `scheduler_fifo_mt()`).